sw/device/lib/dif/dif_kmac.c

```diff
@@ -4,4 +4,478 @@
+   *
+   * The actual number of usable bits may be lower than the value defined
+   * depending on the mode in use. The intent is that this constant is useful
+   * for sizing fixed length buffers.
+   */
+  kDifKmacMaximumBitRate = 1600 - (2 * 128),
+
+  /**
+   * The offset of the second share within the output state register.
+   */
+  kDifKmacStateShareOffset = 0x100,
+};
+
+/**
+ * Report whether the hardware is currently idle.
+ *
+ * If the hardware is not idle then the `CFG` register is locked.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently idle or not.
+ */
+static bool is_state_idle(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_IDLE_BIT);
+}
+
+/**
+ * Report whether the hardware is currently in the absorb state and accepting
+ * writes to the message FIFO.
+ *
+ * Note that writes to the message FIFO may still block if it is full.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently absorbing or not.
+ */
+static bool is_state_absorb(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_ABSORB_BIT);
+}
+
+/**
+ * Report whether the hardware is currently in the squeeze state which means
+ * that the output state is valid and may be read by software.
+ *
+ * @param params Hardware parameters.
+ * @returns Whether the hardware is currently in the squeeze state or not.
+ */
+static bool is_state_squeeze(dif_kmac_params_t params) {
+  uint32_t reg = mmio_region_read32(params.base_addr, KMAC_STATUS_REG_OFFSET);
+  return bitfield_bit32_read(reg, KMAC_STATUS_SHA3_SQUEEZE_BIT);
+}
+
+/**
+ * Wrapper around mmio_region_write32 that reads back the register after
+ * it has been written to verify that it was written successfully.
+ *
+ * @param base_addr Base address.
+ * @param offset Offset of target register.
+ * @param value The value to write to the register.
+ * @returns Whether the register was successfully set to `value`.
+ */
+static bool checked_mmio_region_write32(mmio_region_t base_addr,
+                                        ptrdiff_t offset, uint32_t value) {
+  mmio_region_write32(base_addr, offset, value);
+  return mmio_region_read32(base_addr, offset) == value;
+}
+
+dif_kmac_result_t dif_kmac_init(dif_kmac_params_t params, dif_kmac_t *kmac) {
+  if (kmac == NULL) {
+    return kDifKmacBadArg;
+  }
+
+  *kmac = (dif_kmac_t){.params = params};
+  return kDifKmacOk;
+}
+
+dif_kmac_result_t dif_kmac_configure(dif_kmac_t *kmac,
+                                     dif_kmac_config_t config) {
+  if (kmac == NULL) {
+    return kDifKmacBadArg;
+  }
+
+  // Entropy mode.
+  uint32_t entropy_mode_value;
+  bool entropy_ready = false;
+  switch (config.entropy_mode) {
+    case kDifKmacEntropyModeIdle:
```

Comment:
I don't think the execution time of these small switch statements will be that significant. Are you anticipating DV changing the configuration in a loop without executing other code?

If we do think it will be significant then we should probably split this function into two parts, one to convert from configuration struct to register values and then one to actually set the registers. We could even use a packed configuration struct instead of a function. That way the DV program can simply precalculate and cache the configurations it wants to use.

Reply:
I don't want to block this PR based on this discussion, so we can take this as a separate issue.  I think it would be good to get inputs from DV in this point.

By static_assert I meant that most of these encodings can be directly stored on the enums exposed to the user. I still believe the underlying mmio transactions optimize well into execution.

@cfrantz, @sriyerg, I'd be good to hear your thoughts on this.

Reply:
To the extent possible, we should minimize the amount of indirection.

For cases where there is a 1:1 mapping with register values, I think it makes sense to simply create enumerated constants which map directly to their register bit-patterns and the values can be cast to their correct type (e.g. uint32_t).

For cases where there is a more complex configuration such as a certain named configuration requiring a particular series of register accesses to effect that configuration, then a more traditional enumeration makes sense since the code will need to use switch/case or an if/else tree to set the configuration.

Reply:
I've simplified this a bit by removing some of the enums in the configuration struct and using bools instead. I think this is OK since they aren't being passed directly as arguments (i.e. we don't end up with functions with signatures like `f(bool, bool, bool)`).

> I think it makes sense to simply create enumerated constants which map directly to their register bit-patterns

I think this would require us to pull the generated register file into the DIF header which so far we've avoiding doing. I suggest we revisit doing this for all the DIFs later.

Reply:
SG.